src/realm/object-store/shared_realm.hpp

```diff
@@ -63,22 +64,56 @@ class RealmCoordinator;
 class RealmFriend;
 } // namespace _impl
 
-class Realm : public std::enable_shared_from_this<Realm> {
-public:
-    // A callback function to be called during a migration for Automatic and
-    // Manual schema modes. It is passed a SharedRealm at the version before
-    // the migration, the SharedRealm in the migration, and a mutable reference
-    // to the realm's Schema. Updating the schema with changes made within the
-    // migration function is only required if you wish to use the ObjectStore
-    // functions which take a Schema from within the migration function.
-    using MigrationFunction = std::function<void(SharedRealm old_realm, SharedRealm realm, Schema&)>;
-
-    // A callback function to be called the first time when a schema is created.
-    // It is passed a SharedRealm which is in a write transaction with the schema
-    // initialized. So it is possible to create some initial objects inside the callback
-    // with the given SharedRealm. Those changes will be committed together with the
-    // schema creation in a single transaction.
-    using DataInitializationFunction = std::function<void(SharedRealm realm)>;
+// A callback function to be called during a migration for Automatic and
+// Manual schema modes. It is passed a SharedRealm at the version before
+// the migration, the SharedRealm in the migration, and a mutable reference
+// to the realm's Schema. Updating the schema with changes made within the
+// migration function is only required if you wish to use the ObjectStore
+// functions which take a Schema from within the migration function.
+using MigrationFunction = std::function<void(SharedRealm old_realm, SharedRealm realm, Schema&)>;
+
+// A callback function to be called the first time when a schema is created.
+// It is passed a SharedRealm which is in a write transaction with the schema
+// initialized. So it is possible to create some initial objects inside the callback
+// with the given SharedRealm. Those changes will be committed together with the
+// schema creation in a single transaction.
+using DataInitializationFunction = std::function<void(SharedRealm realm)>;
+
+// A callback function called when opening a SharedRealm when no cached
+// version of this Realm exists. It is passed the total bytes allocated for
+// the file (file size) and the total bytes used by data in the file.
+// Return `true` to indicate that an attempt to compact the file should be made
+// if it is possible to do so.
+// Won't compact the file if another process is accessing it.
+//
+// WARNING / FIXME: compact() should NOT be exposed publicly on Windows
+// because it's not crash safe! It may corrupt your database if something fails
+using ShouldCompactOnLaunchFunction = std::function<bool(uint64_t total_bytes, uint64_t used_bytes)>;
+
+struct RealmConfig {
```

Comment:
RealmConfig and AuditConfig have a circular reference between them, and having it be a nested type inside Realm meant that the cycle couldn't be broken with a forward declaration.